Create etcd image with etcd-backup script

pbitty opened the pull request.

pbitty committed:
Create etcd image with etcd-backup script

jharlap commented on etcd-backup/etcd-backup.py:
```diff
@@ -0,0 +1,164 @@
+    retries = S3_UPLOAD_MAX_RETRIES
+    delay = S3_UPLOAD_RETRY_INITIAL_DELAY_SEC
+
+    while True:
```
just a tip: it would be a lot easier to read the program flow if, instead of using an infinite loop with several ways to break out of it (i.e., a mix of breaks and exceptions), you stated explicitly what the loop conditions are and just modified their state within the loop. currently it took me a few passes to get that this runs until should_shut_down=True OR upload & metrics succeed OR we reach a retry limit. i also had to scroll far away to find out what the retry limit _is_ even thought it's only used here.

jharlap replied:
now i see that this is considered common practice in python. kinda icky to this non-pythonista, but maybe the right thing to do if it's idiomatic for python.

pbitty replied:
I'll rethink this if I can.

jharlap commented on etcd-backup/etcd-backup.py:
```diff
@@ -0,0 +1,164 @@
+            logging.error("Error uploading to S3: %s" % e.message)
+            retries -= 1
+            if retries == 0:
+                raise e
```
you might want to generate logs and/or metrics here - you've given up on a backup and there's zero visibility into that event.

pbitty replied:
In a case like this you get a stack trace as the error bubbles all the way out of the program.  I thought this would be clear enough.  I could wrap it in another error before raising.  Do you have any other suggestions?

jharlap replied:
ooohhh! didn't realize this would crash the program - just thought it'd give up on that backup and try again for the next one.

andrewsykim replied:
LOL, what is this? golang?

jharlap commented on .travis.yml:
```diff
@@ -0,0 +1,11 @@
+  - sudo pip install awscli
+
+script:
+  - sudo make test
```
why do you need sudo to run `make test`?

pbitty replied:
I did this as a quick hack, with the intention of looking it up later.  For some reason the test script fails when trying to create the `tmp` folder.

jharlap replied:
may i introduce you to `mktemp(1)`? ;P

jharlap commented on .travis.yml:
```diff
@@ -0,0 +1,11 @@
+  - docker
+
+before_install:
+  - sudo apt-get install build-essential
```
what do you need that for? awscli?

jharlap replied:
if you had said this was a python project, would it not already have a build toolchain and pip ready to use?

jharlap commented on .travis.yml:
```diff
@@ -0,0 +1,11 @@
+
+before_install:
+  - sudo apt-get install build-essential
+  - sudo curl -L "https://github.com/docker/compose/releases/download/1.9.0/docker-compose-$(uname -s)-$(uname -m)" -o /usr/local/bin/docker-compose
```
they don't have docker-compose pre-installed?

jharlap reviewed (changes requested):
- not entirely clear why this needs to package both etcd and the backup script in the same image. it seems like any etcd image would work with this backup tool so long as you volume mount the data dir.
- the backup script never cleans up after itself, so eventually disk space within the container will fill up.
- there is logging & metrics for the happy code path, but not so much for the unhappy code paths - you're going to appreciate some logging/metrics there when it's time to debug the running app.
- if you're using sudo, i think you need to specify such in the travis.yml.
- you might want to include a suggestion in the readme about using s3 lifecycle policies to purge old backups
- how do you ensure your tests run in the right order? specifically, how do you ensure all the etcd containers are up and ready before you get to `run_tests()`?

pbitty committed:
Try Python environment in Travis

pbitty committed:
Try build w/o build-essential

pbitty committed:
Try w/o installing docker-compose

andrewsykim commented on Dockerfile:
```diff
@@ -0,0 +1,23 @@
+ENV ETCD_BUILD="etcd-v${ETCD_VERSION}-linux-amd64"
+
+RUN apk --no-cache --virtual build-deps add openssl && \
+    wget "https://github.com/coreos/etcd/releases/download/v2.3.7/${ETCD_BUILD}.tar.gz" -O - | \
```
should v2.3.7 also be using ${ETCD_VERSION}?

andrewsykim commented on README.md:
````diff
@@ -1,3 +1,34 @@
+
+    ```shell
+    docker run \
+      -e ETCD_LISTEN_PEER_URLS=http://0.0.0.0:2380 \
````
Wouldn't you need hostNetwork to listen at 0.0.0.0? Otherwise the other members can't reach you.

andrewsykim commented on etcd-backup/etcd-backup.py:
```diff
@@ -0,0 +1,164 @@
+import sys
+
+BACKUP_BASE_DIR = "/tmp"
+LOG_LEVEL = os.getenv("LOG_LEVEL", "INFO")
```
Not sure if this is idiomatic python, if not should include in the docs?

andrewsykim commented on etcd-backup/etcd-backup.py:
```diff
@@ -0,0 +1,164 @@
+    except subprocess.CalledProcessError as e:
+        logging.error("Error running etcdctl. Output: %s" % e.output)
+        raise
+    return backup_dir
```
Why do we need to return this?

andrewsykim commented on etcd-backup/etcd-backup.py:
```diff
@@ -0,0 +1,164 @@
+    logging.debug("Compressing backup directory %s into tarball %s" % (source_dir, file_path))
+    with tarfile.open(name=file_path, mode="w:gz") as tar:
+        tar.add(source_dir, arcname='/')
+    return file_path
```
why do we need to return file_path here?

andrewsykim commented on etcd-backup/etcd-backup.py:
```diff
@@ -0,0 +1,164 @@
+
+
+def submit_metrics(bucket, prefix, file_size_bytes):
+    api_key = os.getenv('DATADOG_API_KEY')
```
We should mention something about it in the README?

andrewsykim replied:
that you can optionally use datadog to send metrics that is

andrewsykim reviewed (changes requested):
Not done reviewing, laptop just running out of batteries! Will review again when I'm in.

pbitty committed:
Remove unneeded return values

pbitty committed:
A documentation on Datadog support

pbitty committed:
Add LOG_LEVEL to usage info

pbitty committed:
Fix etcd download URL

pbitty commented:
@jharlap  Regarding logging the unhappy code paths, I think we're ok with letting exceptions bubble up.

For metrics in the unhappy code paths, I was thinking of trying a different approach.  Instead of caring about when errors happen, we care about the absence of a success.

So we'd alert on the _lack_ of backup metrics, instead of looking for failed backups.  I actually think we should do the same for our current DB and Redis backups, because we are actually open to the situation where the failure metric reporting doesn't work (it has happened before), and we were are then unaware that a backup isn't happening.

What do you think?

pbitty committed:
Fix method clobbered by docstring

pbitty committed:
Clean up after backup

pbitty committed:
Remove unnecessary infinite loop

Explain reasoning for remaining infinite loop

pbitty committed:
Re-order Dockerfile for better build caching

pbitty commented on etcd-backup/etcd-backup.py:
```diff
@@ -0,0 +1,177 @@
+
+        # We want to exit before sleeping, so we check this here, not as the loop condition
+        if should_shut_down:
+            break
```
@jharlap, this is why I had to use the infinite loop.

pbitty committed:
Move etcd-backup to repo root

pbitty committed:
Keep only etcd-backup functionality, remove etcd from image

Use CoreOS etcd image for tests

pbitty committed:
Wait for endpoints to be ready before proceeding in tests

pbitty commented:
Changes addressed.  R please?

pbitty committed:
Update README

pbitty committed:
Add note about test requiring sudo

pbitty commented:
@jharlap  I tried using `mktemp` to avoid having to run the tests under `sudo`, but it still had issues cleaning up because the files are created inside a container, by procs running as root.

It works on my machine I believe because I'm running Docker for Mac, which handles the permissions translation and keeps host-level permissions separate from container-level permissions.  Anyway, I'm going to leave as is for now, and re-visit if it becomes an issue.

pbitty commented:
So... I forgot an important point, we only want this running on one host at a time.   So we need some way to lock.  [This person](https://crewjam.com/etcd-aws/) just checked if we were running on the leader and only did the backup on the leader.   We really could just be using his code, but I wasn't confident in his process of just scanning the tree.

Anyway, I'll push more commits soon.

pbitty committed:
Only run backup on leader

pbitty committed:
Use dogstatsd instead of Datadog HTTP API

We want to be able to link these metrics to the running host, which
dogstatsd will happily do for us.

andrewsykim commented on README.md:
```diff
@@ -1,3 +1,33 @@
 # docker-etcd
```
update this?

pbitty committed:
Update title in README

andrewsykim commented on README.md:
```diff
@@ -1,3 +1,33 @@
+
+See the `print_usage` method in [etcd-backup.py](etcd-backup.) for all available environment variables.
+
+## Leader Backups
```
You probably considered this already, what happens when leader election changes during backup?

pbitty replied:
We'll eventually have a snapshot from the latest leader.  I'm not too concerned.  Backups will always be a bit stale.  I'm more concerned with them being consistent.

pbitty replied:
I'm not sure if the following is possible, as I don't know exactly how etcd implements Raft, but:

It may be possible in a network partition that a peer thinks it is the leader when it is not.  In that case we may have uploads from two "leaders".  We'd just have to figure out which one is the real one.  I don't think it's worth digging into it too much at this point.

jharlap replied:
I agree that you don't need to sweat it. Stop reading here if you don't want the details ;)

I don't think it is a question of how _etcd_ implements Raft - it's just a property of Raft that for a short time after a network fault (i.e., until the heartbeat interval ends) a former leader who is now in a minority partition will still believe it is the leader. However, writes are only _committed_ after a quorum of nodes have stored the write, so even though the leader is still a leader, it will never manage to get a quorum of writes so it will not commit the change and thus the cluster is unavailable during this time period. Once the heartbeat interval fires, a new leader is elected within the majority partition (assuming the majority is enough to achieve quorum) and then the cluster becomes available again (assumption: clients can reach the new leader).

andrewsykim replied:
I always want the details :P

pbitty replied:
Me too.  :)  @jharlap That makes sense to me.

I'm more curious about the leader left behind in the minority partition.  It won't be able to commit writes as you said, but elections will also fail in the minority partition.  In that case, what does the node report as its `state`?

Not sweating it for now.  Bigger 🌽 to plant and harvest.

pbitty commented:
@andrewsykim @jharlap Ready for review again.

jharlap reviewed (changes requested):
your test suite will print out success vs failure, but the exit code of the test is always 0 so travis will think it passes even when it fails

pbitty committed:
Exit with non-zero code when tests fail

pbitty commented:
Silly me.  Test fixed.  R?

pbitty committed:
Push to registry on master merge builds
